# CLASE 12: FUNCIONES LINEALES

# 5. Funciones Lineales

La notación $f: \mathbb{R}^{n}\to \mathbb{R}$ significa que $f$ es una función con **dominio** en $\mathbb{R}^{n}$ e imagen en $\mathbb{R}$, es decir, que toma un $n$-vector y lo convierte en un escalar.
  
La función puede denotarse como $f\left(\mathbf{x}\right)= f\left(x_{0},x_{1},\dots,x_{n-1}\right)$ donde $\mathbf{x}\in \mathbb{R}^{n}$ y $x_{i}$ son las entradas del vector.  

Es necesario enfatizar que $f(\mathbf{x})$ es un escalar. 

Para describir una función $f: \mathbb{R}^{n} \to \mathbb{R}$ tenemos que especificar el cual es el valor para cualquier vector $\mathbf{x}\in\mathbb{R}^{n}$. Por ejemplo, sea $f: \mathbb{R}^{4} \to \mathbb{R}$ definida por
$$
f(\mathbf{x}) = x_{0} + x_{1} + x_{2} + x_{3}
$$

Así pues el valor de cualquier $4$-vector queda bien definido, por ejemplo:

* $f(\mathbf{0}) = 0 + 0+ 0+0 = 0$

* $f([1\; 0\; 0\; 0]) = 1+0+0+0 = 1$

* $f([1\; 1\; 1\; 1]) = 1+1+1+1 = 4$

In [3]:
# Se llama las librerías necesarias
import numpy as np

In [6]:
# Función para crear la función que aparece anteriormente
def f(x):
    # np.sum realiza el x[0] + x[1] + x[2] + x[3]
    return np.sum(x)

In [7]:
# Se evaluan diferentes vectores
print('f(0):\n', f(np.array([0,0,0,0])))
print('f([1,0,0,0]):\n', f(np.array([1,0,0,0])))
print('f([1,1,1,1]):\n', f(np.array([1,1,1,1])))

f(0):
 0
f([1,0,0,0]):
 1
f([1,1,1,1]):
 4


Como otro ejemplo, sea $g: \mathbb{R}^{3} \to \mathbb{R}$ definida por 
$$
g(\mathbf{x}) = x_{0}
$$

Tenemos entonces:

* $g(\mathbf{0}) = 0$ 

* $g([0\; 0\; 1]) = 0$

* $g([1\; 0\; 4]) = 1$

In [8]:
# Se crea la función g
def g(x):
    return x[0]

In [9]:
print('g(0):\n', g(np.array([0,0,0])))
print('g([0,0,1]):\n', g(np.array([0,0,1])))
print('g([1,0,4]):\n', g(np.array([1,0,4])))

g(0):
 0
g([0,0,1]):
 0
g([1,0,4]):
 1


Podemos notar que en este caso $f(\mathbf{x})$ corresponde a la _función suma_ y $g(\mathbf{x})$ a proyección sobre $x_{0}$.

# CLASE 13: ALGUNAS FUNCIONES LINEALES

## 5.1 Algunas funciones

**Función producto interno**. Si una función tiene como aurgmentos dos $n$-vectores el dominio de la función se expresará como $\mathbb{R}^{n}\times\mathbb{R}^{n}$. Por ejemplo, sea $f:\mathbb{R}^{n}\times\mathbb{R}^{n} \to \mathbb{R}$ definida por:

$$
f(\mathbf{a},\mathbf{b}) = a_{0}b_{0} + a_{1}b_{1} + \cdots + a_{n-1}b_{n-1}
$$

Notemos que podemos expresar lo anterior como

$$
f(\mathbf{a},\mathbf{b}) = \mathbf{a}^{T}\cdot \mathbf{b}
$$

Esta función recibe el nombre de función producto interno. Si quisieramos dejar fijo un vector, digamos el vector $\mathbf{a}$ podemos recurrir a la siguiente notación:

$$
f_{\mathbf{a}}(\mathbf{b}) = \mathbf{a}^{T}\mathbf{b}
$$

Esto significará que $\mathbf{a}$ estará fijo mientras que $\mathbf{b}$ será el que variará. 

**Superposición y linealidad.** La función producto interno satisface la siguiente propiedad:

$$
\begin{matrix}
f_\mathbf{x}(\alpha\mathbf{a}+\beta\mathbf{b})  & = & \mathbf{x}^{T}(\alpha\mathbf{a}+\beta\mathbf{b}) \\ 
& = & \mathbf{x}^{T}(\alpha\mathbf{a})+\mathbf{x}^{T}(\beta\mathbf{b})\\
 & = & \alpha(\mathbf{x}^{T}\mathbf{a}) + \beta(\mathbf{x}^{T}\mathbf{b})\\
 & = & \alpha f_{\mathbf{x}}(\mathbf{a}) + \beta f_{\mathbf{x}}(\mathbf{b}) 
\end{matrix}
$$

Si la función $g$ cumple que $g(\alpha\mathbf{x}+\beta\mathbf{y}) = \alpha g(\mathbf{x})+\beta g(\mathbf{y})$ decimos que la función $g$ es una **función lineal.** 

Si $f$ es una función lineal la _linearidad_ se extiende a cualquier combinación lineal y no solamente a una combinación de dos vectores, tenemos 

$$
f(\alpha_{0}\mathbf{x_{0}} + \alpha_{1}\mathbf{x_{1}} + \alpha_{2}\mathbf{x_{2}} + \cdots 
 + \alpha_{n-1}\mathbf{x_{n-1}}) = \alpha_{0}f(\mathbf{x_{0}}) + \alpha_{1}f(\mathbf{x_{1}}) + \alpha_{2}f(\mathbf{x_{2}}) + \cdots + \alpha_{n-1}f(\mathbf{x_{n-1}})
$$


Si una función $g:\mathbf{R}^{n} \to \mathbf{R}$ es lineal sabemo que cumple $g(\alpha\mathbf{x}+\beta\mathbf{y}) = \alpha g(\mathbf{x})+\beta g(\mathbf{y})$, propiedad que llamaremos _superposición_. Ahora, la superposición puede ser separada en dos propiedades:

* _Homogeneidad_. Para cualquier $n$-vector $\mathbf{x}$ y cualquier escalar $\alpha$ tenemos que $g(\alpha \mathbf{x}) = \alpha g(\mathbf{x})$

* _Aditividad_. Para cualquier $n$-vector $\mathbf{x}$ y $\mathbf{y}$ tenemos que $g(\mathbf{x}+\mathbf{y}) = g(\mathbf{x})+g(\mathbf{y})$

La homogeneidad nos dice que escalar el argumento es igual que escalar el valor de la función; la aditividad nos indica que sumar los argumentos es igual que sumar los valores de la función. 

In [10]:
# Se comprueba si la función f creada anteriormente cumple con la
# propiedad de la homogeneidad
# Se crean los vectores
a = np.array([1,1,1,1])
b = np.array([1,0,1,0])
# Se crean los escalares
x,y = 1,-2

In [11]:
print('x*a + y*b:\n', x*a + y*b)
print('f(x*a + y*b):\n', f(x*a + y*b))
print('x*f(a) + y*f(b):\n', x*f(a) + y*f(b))

x*a + y*b:
 [-1  1 -1  1]
f(x*a + y*b):
 0
x*f(a) + y*f(b):
 0


Se comprueba que la función cumple con la propiedad de la homogeneidad, por lo que es lineal.

# CLASE 14: UN TEOREMA EN FUNCIONES LINEALES

## 5.1 Algunas funciones

**Representación producto interno de una función lineal.** En matemáticas las verdades, cuando son verdades, se demuestran usando argumentos principalmente lógicos. Estas verdades vienen empaquetadas en _teoremas_, _corolarios_ y _lemas_. Si no podemos demostrarlo pero creemos que es verdad lo llamamos _conjetura_. Por ejemplo, consideremos el siguiente teorema:

**Teorema.** Los ángulos a ambos lados de una recta, que intersecta a otra, siempre suman $180^\circ$ 

![](imagenes/img3.png)

*Demostración*: Supongamos que $\alpha + \beta \neq 180$ eso quiere decir que $\alpha + \beta > 180$ o $\alpha + \beta < 180$. Ahora, pensemos en la linea **f** pasa por el diámetro, es decir, corta a la mitad al círculo. 

![](imagenes/img4.png)

Al pasar por el diámetro sabemos que ambos lados tenemos un ángulo de $180^\circ$. Llamemos a ese ángulo $\Omega=180^\circ$. Ahora, llamemos **g** a una línea que pase por otro diámetro. 

![](imagenes/img5.png)

Entonces, al seccionar con **g** tenemos nuevamente los ángulos $\alpha$ y $\beta$ entonces tenemos que $\alpha + \beta = \Omega$ pero asumimos que $\alpha + \beta \neq 180^\circ$ lo cuyal nos lleva a una contradicción $!$ por ello lo anteior debe ser cierto.  

**Corolario.** Cuando dos líneas se intersectan los ángulos opuestos entre ellos son iguales.

![](imagenes/img6.png)

*Demostración:* Sabemos que $\beta + \alpha = 180^\circ$ entonces $\alpha = 180^\circ - \beta$. Por otro lado, sabemos que también $\gamma + \beta = 180^\circ$ así pues $\gamma = 180^\circ - \beta$. Juntando lo anteior tenemos que:

$$
\alpha  =  180^\circ - \beta = \gamma
$$

Lo mismo es posible aplicarse para $\beta$ y $\delta$. **QED**. 

Ya explorados los teoremas y los corolarios veamos la siguiente proposición:  

**Proposición:** Una función definida como el producto punto de sus argumentos y algún vector fijo es lineal. Lo inverso es también verdadero: si una función es lineal entonces puede ser expresada como el producto interno de sus argumentos y algún vector fijo.

Ahora, sea $f$ una función escalar de $n$-vectores que además es lineal. Entonces decimos que _existe_ un $n$-vector $\mathbf{a}$ tal que $f(\mathbf{x}) = \mathbf{a}^{T}\mathbf{x}$ para cualquier $\mathbf{x}$. A la expresión $\mathbf{a}^{T}\mathbf{x}$ le llamaremos *representación producto punto* de $f$. 

Otra propiedad interesante es que la *representación producto punto* de $f$ como $\mathbf{a}^{T}\mathbf{x}$ es única. 

$ f(\mathbf{x}) = \displaystyle\sum_{i}^{n-1} x_{i}$ si n=4 esto lo podemos expresar como $f(\mathbf{x}) = \mathbf{a}^{T}\mathbf{x}$ donde:
$$
a^T = [1,1,1,1]
$$

**Promedio.** El promedio de una $n$-vector es definida como

$$
f(x) = (x_{0}+x_{1}+\cdots+x_{n-1})/n
$$

Y se denota como $\mathbf{avg}(x)$, notemos que este $\mathbf{avg}(x) = a^{T}x$ con

$$
a^T = [1/n\; 1/n\; \cdots \; 1/n] = \mathbf{1}/n
$$

**Máximo.** Para $n$-vectores $x$ la función que nos da el máximo coeficiente de un vector es $f(x)=max\{x_{0},x_{1},\dots,x_{n-1}\}$ y no es lineal. Para demostrar esto voy a suponer que es lineal y dar un contraejemplo. Si tomamos $x^T = [1,-1]$ y $y^T=[-1,1]$ con $\alpha = 1/2$ y $\beta = 1/2$. Entonces tenemos que:

$$
f(\alpha x + \beta y) = 0 \neq \alpha f(x) + \beta f(y) = 1
$$

In [12]:
# Se comprueba la linealidad de la función máximo
# función para hallar el número maximo
def MAX(x):
    return np.max(x)

In [13]:
print('MAX([1,0,2]):\n', MAX(np.array([1,0,2])))

MAX([1,0,2]):
 2


La función creada funciona correctamente

In [15]:
# Se crean los vectores
x = np.array([1,-1])
y = np.array([-1,1])
# Se crean los escalares
alpha, beta = 1/2, 1/2

In [18]:
print('MAX(alpha*x + beta*y):\n', MAX(alpha*x + beta*y))
print('MAX(alpha*x) + MAX(beta*y):\n', MAX(alpha*x) + MAX(beta*y))


MAX(alpha*x + beta*y):
 0.0
MAX(alpha*x) + MAX(beta*y):
 1.0


La función MAX no cumple con la propiedad de la superposición, por tanto no es una función líneal.

# CLASE 15: FUNCIONES AFINES

## 5.2 Algunas consideraciones

**Funciones Afines.** Una función $f:\mathbb{R}^{n} \to \mathbb{R}$ es afín si y solo si puede ser expresada como $f(x) = a^{T}x + b$ para algún $n$-vector $a$ y un escalar $b$, es llamado algunas veces *offset* o *bias*. Por ejemplo si consideramos la siguiente función:

$$
f(x) = 3 + x_{0}+x_{1}+x_{2}
$$

La podemos representar como 

$$
f(x) = a^T x + b = [1 \; 1 \; 1] x + 3
$$

Una función afín cumplirá la superposición si y solo si la suma de de los coeficientes de la combinación lineal es uno. Es decir, si $f$ es una función afín entonces $f(\alpha x + \beta y) = \alpha f(x)+ \beta f(y)$ si y solo si $  \alpha + \beta = 1$.

*Demostración.* Como $f$ es afín tenemos que $f(x)=a^T x + b$  entonces, usando que $  \alpha + \beta = 1$

$$
\begin{matrix}
f(\alpha x + \beta y) &=& a^T (\alpha x + \beta y) + b \\
 &=& a^T \alpha x + a^T \beta y + (\alpha + \beta) b \\
 &=& \alpha a^T x + \alpha b + \beta a^T y + \beta b \\
 &=& \alpha (a^T x + b) + \beta (a^T y + b) \\
 &=& \alpha f(x) + \beta f(y)
\end{matrix}
$$

**Ejercicio.** Determina si las siguientes funciones son lineales o no:
* $f(\mathbf{x}) = x_{n-1} - x_{0}$

**Solución**

In [1]:
# Se crea la función
def fa(x):
    n = len(x)
    return x[n-1] - x[0]

In [8]:
# Se crean los vectores
x = np.array([1,2,3,4])
y = np.array([5,6,7,8])
# Se crean los escalares
alpha = 1
beta = 2

In [9]:
# Se comprueba las función 
print('f(alpha*x + beta*y):\n', fa(alpha*x + beta*y))
print('f(alpha*x) + f(beta*y):\n', fa(alpha*x) + fa(beta*y))

f(alpha*x + beta*y):
 9
f(alpha*x) + f(beta*y):
 9


La función cumple con la propiedad de la superposición, por tanto es una función lineal

**Ejercicio.** Determina si las siguientes funciones son lineales o no:

* Si $\mathbf{x}\in\mathbb{R}^{4}$ y $f:\mathbb{R}^{4} \to \mathbb{R}$ definida como $f(\mathbf{x})=x_{3}+(x_{3}-x_{2})$

**Solución**


In [10]:
# Se crea la función
def fb(x):
    n = len(x)
    if n == 4:
        return x[3] + (x[3] - x[2])
    else:
        print('Se requiere vectores en R4')

In [11]:
# Se crean los vectores
x = np.array([1,2,3,4])
y = np.array([5,6,7,8])
# Se crean los escalares
alpha = 1
beta = 2

In [13]:
# Se comprueba las función 
print('f(alpha*x + beta*y):\n', fb(alpha*x + beta*y))
print('f(alpha*x) + f(beta*y):\n', fb(alpha*x) + fa(beta*y))

f(alpha*x + beta*y):
 23
f(alpha*x) + f(beta*y):
 11


La función no cumple con la propiedad de la superposición, por tanto no es una función lineal.

**Problema: Potencia en un clúster de procesadores y temperatura.** La temperatura $T$ de un cluster de tres procesadores es una función afín de la potencia de los procesadores $P^T = [P_{0}\; P_{1}\; P_{2}]$. Cuando los tres procesadores están en reposo tenemos que $P^T = [10\; 10\; 10]$ lo cual nos da una temperatura de $T=35$. Cuando el primer procesador opera a potencia absoluta tenemos que $P^T = [100\; 10\; 10]$ y la temperatura se eleva a $T=60$. Cuando el segundo procesador opera a su máxima capacidad tenemos $P^T = [10\; 100\; 10]$ y la temperatura se eleva a $T=75$. Cuando el tercer procesador opera a máxima potencia mientras que los otros dos están en reposo tenemos que $P^T = [10\; 10\; 100]$ y la temperatura se elevea a $T=65$. Supongamos que los tres procesadores operan a máxima potencia ¿Qué temperatura se alcanza? La temperatura de fusión de los procesadores (a la que los procesadores se derriten) es cerca de $150$ ¿Cuál es la potencia máxima que pueden alcanzar los procesadores para que $T\le 150$?  
*Hint:* $T(P) = a^T P + b$

**Solución**

Primero se escribe un sistema de ecuaciones que represente la ecuación $T(P) = a^T P + b$, con los valores dados en el ejemplo:

$$
\begin{matrix}
10a_0 + 10a_1 + 10a_2 + b &=& 35 \\ 
100a_0 + 10a_1 + 10a_2 + b &=& 60 \\ 
10a_0 + 100a_1 + 10a_2 + b &=& 75 \\ 
10a_0 + 10a_1 + 100a_2 + b &=& 65 
\end{matrix}
$$

Se tiene un sistema de 4 ecuaciones con 4 incognitas, por tanto se puede escribir de forma matricial (Tambien se puede realizar despejando variablese) de la siguiente forma.

$$
\begin{bmatrix}
10 & 10 & 10 & 1 \\ 
100 & 10 & 10 & 1 \\ 
10 & 100 & 10 & 1 \\ 
10 & 10 & 100 & 1 
\end{bmatrix}\cdot
\begin{bmatrix}
a_0 \\ a_1 \\ a_2 \\ b
\end{bmatrix}=
\begin{bmatrix}
35 \\ 60 \\ 75 \\ 65
\end{bmatrix}
$$

A la matriz de 4x4 se le llamara $X$, al vector con las incognitas como $const$ y al vector con las soluciones como $sol$, se escribe esto en python.

In [36]:
# Matriz
X = np.array([[10,10,10,1],
              [100,10,10,1],
              [10,100,10,1],
              [10,10,100,1]])

sol = np.array([35,60,75,65])

print('X:\n', X)
print('sol:\n', sol)

X:
 [[ 10  10  10   1]
 [100  10  10   1]
 [ 10 100  10   1]
 [ 10  10 100   1]]
sol:
 [35 60 75 65]


Por tanto se tiene un sistema:

$$
X\cdot const = sol
$$

Se le aplica la inversa de X a ambos lados, cancelando la X en el lado izquierdo, dejando solo el vector con las incognitas

$$
X^{-1}\cdot X\cdot const = X^{-1}\cdot sol\newline
const = X^{-1}\cdot sol
$$

En python esto queda como:


In [47]:
# Inversa de X
X_inv = np.linalg.inv(X)
# Se halla los valores de la incognitas
const = X_inv.dot(sol)
print('const:\n', const)

const:
 [ 0.27777778  0.44444444  0.33333333 24.44444444]


La función para hallar la temperatura según la potencia de los procesadores esta dada por

$$
T(P) = a^T P + b
$$

El vector hallado $const$ esta compuesto por 

$$
const = [a_0, a_1, a_2, b]
$$

El vector a esta compuesto por los primeros tres elemento y b por el último elemento, por tanto, escribimos esto en python


In [50]:
a = const[0:3]
b = const[3]
print('a:\n', a)
print('b:\n', b)


a:
 [0.27777778 0.44444444 0.33333333]
b:
 24.44444444444444


Con los vectores $a$ y $b$, ya se tiene los vectores constantes para hallar la temperatura solamente falta crear la función que depende únicamente del vector de la potencia de los procesadores 

$$
T(P) = a^T P + b
$$

In [51]:
def temp(a, b, P):
    return a.T.dot(P) + b

Se comprueba la temperaturas con los valores de la potencias dadas en el enunciado

In [54]:
print('T(P) para P[10,10,10]:\n', temp(a, b, np.array([10,10,10])))
print('T(P) para P[100,10,10]:\n', temp(a, b, np.array([100,10,10])))
print('T(P) para P[10,100,10]:\n', temp(a, b, np.array([10,100,10])))
print('T(P) para P[10,10,100]:\n', temp(a, b, np.array([10,10,100])))

T(P) para P[10,10,10]:
 35.0
T(P) para P[100,10,10]:
 60.000000000000014
T(P) para P[10,100,10]:
 75.00000000000001
T(P) para P[10,10,100]:
 65.0


Se puede observar que el sistema funciona correctamente.

Supongamos que los tres procesadores operan a máxima potencia ¿Qué temperatura se alcanza? 

In [55]:
print('T(P) para P[100,100,100]:\n', temp(a, b, np.array([100,100,100])))

T(P) para P[100,100,100]:
 130.0


A máxima potencia se alcanza los 130°

La temperatura de fusión de los procesadores (a la que los procesadores se derriten) es cerca de $150$ ¿Cuál es la potencia máxima que pueden alcanzar los procesadores para que $T\le 150$? 

In [58]:
print('T(P) para P[120,120,120]:\n', temp(a, b, np.array([120,120,120])))

T(P) para P[120,120,120]:
 151.11111111111111


La potencia máxima para derretir los procesadores es de 120.